src/main/java/soot/AbstractSootFieldRef.java

```diff
@@ -97,42 +95,51 @@ public FieldResolutionFailedException() {
 
     @Override
     public String toString() {
-      StringBuffer ret = new StringBuffer();
-      ret.append(super.toString());
+      StringBuilder ret = new StringBuilder(super.toString());
       resolve(ret);
       return ret.toString();
     }
   }
 
   @Override
   public SootField resolve() {
-    return resolve(null);
+    SootField cached = this.resolveCache;
+    // Use the cached SootField if available and still valid
+    if (cached == null || !isValidResolve(cached)) {
+      cached = resolve(null);
+      this.resolveCache = cached;
+    }
+    return cached;
+  }
+
+  private boolean isValidResolve(SootField f) {
```

Comment:
Is there a case this function would return false for a cached SootField?

Reply:
The ```isValidResolve(SootField)``` function can return false if the ```SootField``` instance was modified after it was cached here. The ```SootField``` class has methods such as ```setName(..)```, ```setType(..)```, etc. that allow it to be modified.

Reply:
@tim-hoffman this is probably bad. If someone doesn't take care and add more setters to SootField in future without modifying this function. IMHO, it is better to compare hashcode of objects or use an oberser pattern.

Reply:
It's important to note that adding more setters to ```SootField``` in the future could only cause an issue if they overlap with concepts/values which have getters in the ```SootFieldRef``` interface. Witht that said, here are my thoughts on those two options as well as a third which I think provides the best solution.

Option 1 - Observer: ```SootField``` would need to keep a Set of ```AbstractSootFieldRef``` that have previously been resolved to that ```SootField``` and any time the ```SootField``` is modified (in a way that overlaps with the ```SootFieldRef``` interface), the Set must be cleared. ```AbstractSootFieldRef#resolve()``` would need to register itself with that Set in the respective resolved ```SootField```. Then the ```isValidResolve``` check would be replaced with checking if the cached ```SootField``` contains the current ```AbstractSootFieldRef``` instance in its Set. I don't think this is a great solution because it has additional storage overhead for the Set and if someone adds another setter to ```SootField```, it still has the potential of getting out of sync. There is also the potential of someone adding a setter to ```SootField``` and clearing the Set unnecessarily if that new setter does not overlap with the ```SootFieldRef``` interface (that person would have to remember that not all modifications to a ```SootField``` object make a ```SootFieldRef``` for it invalid).
Option 2 - hashCode: This would require adding an additional field alongside the cached ```SootField``` reference to store the hashcode of that reference. However, ```SootField``` does not implement ```#hashCode()``` (which may be intentional so I'd rather not change that). It does have ```#equivHashCode()``` but that does not check the declaring class field (and again, probably intentional so I'd rather not change it). Thus, we either need an ad-hoc solution that combines ```#equivHashCode()``` with the declaring class reference's hashcode or have to introduce another form of hashcode method in ```SootField```. Neither seems great, and again, there is additional memory overhead for the 'int' field to store alongside the cached ```SootField``` reference in ```AbstractSootFieldRef```.
Option 3 - Move ```#isValidResolve(..)``` to the ```SootField``` class and change its signature to ```isValidResolve(SootFieldRef)```. I think this is the best solution because it adds no additional memory overhead for the field/fieldref. I think placing that method in the ```SootField``` class can help with your concern of someone updating the ```SootField``` class by adding something else to it while still making it clear that only additions/changes that also overlap with the ```SootFieldRef``` interface need to used in the comparison.

Reply:
@tim-hoffman  Thanks for elaborating the options, option 3 looks good.

Reply:
Okay, I will make that change and update.